code/modules/atmospherics/machinery/air_alarm/_air_alarm.dm

```diff
@@ -0,0 +1,529 @@
+
+	update_appearance()
+
+	return TRUE
+
+/obj/machinery/airalarm/update_appearance(updates)
+	. = ..()
+
+	if(panel_open || (machine_stat & (NOPOWER|BROKEN)) || shorted)
+		set_light(0)
+		return
+
+	var/area/our_area = get_area(src)
+	var/color
+	switch(max(danger_level, !!our_area.active_alarms[ALARM_ATMOS]))
+		if(0)
+			color = "#03A728" // green
+		if(1)
+			color = "#EC8B2F" // yellow
+		if(2)
+			color = "#DA0205" // red
+
+	set_light(1.4, 1, color)
+
+/obj/machinery/airalarm/update_icon_state()
+	if(panel_open)
+		switch(buildstage)
+			if(AIRALARM_BUILD_COMPLETE)
+				icon_state = "alarmx"
+			if(AIRALARM_BUILD_NO_WIRES)
+				icon_state = "alarm_b2"
+			if(AIRALARM_BUILD_NO_CIRCUIT)
+				icon_state = "alarm_b1"
+		return ..()
+
+	icon_state = "alarmp"
+	return ..()
+
+/obj/machinery/airalarm/update_overlays()
+	. = ..()
+
+	if(panel_open || (machine_stat & (NOPOWER|BROKEN)) || shorted)
+		return
+
+	var/area/our_area = get_area(src)
+	var/state
+	switch(max(danger_level, !!our_area.active_alarms[ALARM_ATMOS]))
+		if(0)
+			state = "alarm0"
+		if(1)
+			state = "alarm2" //yes, alarm2 is yellow alarm
+		if(2)
+			state = "alarm1"
+
+	. += mutable_appearance(icon, state)
+	. += emissive_appearance(icon, state, src, alpha = src.alpha)
+
+/**
+ * main proc for throwing a shitfit if the air isnt right.
+ * goes into warning mode if gas parameters are beyond the tlv warning bounds, goes into hazard mode if gas parameters are beyond tlv hazard bounds
+ *
+ */
+/obj/machinery/airalarm/proc/check_air_dangerlevel(turf/location, datum/gas_mixture/environment, exposed_temperature)
+	SIGNAL_HANDLER
+	if((machine_stat & (NOPOWER|BROKEN)) || shorted)
+		return
+
+	var/datum/tlv/current_tlv
+	//cache for sanic speed (lists are references anyways)
+	var/list/cached_tlv = tlv_collection
+
+	var/list/env_gases = environment.gases
+	var/partial_pressure = R_IDEAL_GAS_EQUATION * exposed_temperature / environment.volume
+
+	current_tlv = cached_tlv["pressure"]
+	var/environment_pressure = environment.return_pressure()
+	var/pressure_dangerlevel = current_tlv.check_value(environment_pressure)
+
+	current_tlv = cached_tlv["temperature"]
+	var/temperature_dangerlevel = current_tlv.check_value(exposed_temperature)
+
+	var/gas_dangerlevel = 0
+	for(var/gas_id in env_gases)
+		if(!(gas_id in cached_tlv)) // We're not interested in this gas, it seems.
+			continue
+		current_tlv = cached_tlv[gas_id]
+		gas_dangerlevel = max(gas_dangerlevel, current_tlv.check_value(env_gases[gas_id][MOLES] * partial_pressure))
+
+	environment.garbage_collect()
+
+	var/old_danger_level = danger_level
+	danger_level = max(pressure_dangerlevel, temperature_dangerlevel, gas_dangerlevel)
+
+	if(old_danger_level != danger_level)
+		INVOKE_ASYNC(src, PROC_REF(apply_danger_level))
+	if(istype(selected_mode, /datum/air_alarm_mode/cycle) && environment_pressure < ONE_ATMOSPHERE * 0.05)
+		var/datum/air_alarm_mode/cycle/typed_mode = selected_mode
+		typed_mode.replace(my_area)
```

Comment:
what is replaced for? why not have this logic inside the mode itself, in a base replaced() proc

Reply:
yeah this one kinda sucks ill figure something sane later.